specs/predeploys.md

```diff
@@ -0,0 +1,670 @@
+     * @param _data Unsigned RLP encoded tx, 6 elements
+     * @return L1 fee that should be paid for the tx
+     */
+    function getL1Fee(bytes memory _data) public view returns (uint256);
+
+    /**
+     * Computes the amount of L1 gas used for a transaction
+     * The overhead represents the per batch gas overhead of
+     * posting both transaction and state roots to L1 given larger
+     * batch sizes.
+     * 4 gas for 0 byte
+     * https://github.com/ethereum/go-ethereum/blob/9ada4a2e2c415e6b0b51c50e901336872e028872/params/protocol_params.go#L33
+     * 16 gas for non zero byte
+     * https://github.com/ethereum/go-ethereum/blob/9ada4a2e2c415e6b0b51c50e901336872e028872/params/protocol_params.go#L87
+     * This will need to be updated if calldata gas prices change
+     * Account for the transaction being unsigned
+     * Padding is added to account for lack of signature on transaction
+     * 1 byte for RLP V prefix
+     * 1 byte for V
+     * 1 byte for RLP R prefix
+     * 32 bytes for R
+     * 1 byte for RLP S prefix
+     * 32 bytes for S
+     * Total: 68 bytes of padding
+     * @param _data Unsigned RLP encoded tx, 6 elements
+     * @return Amount of L1 gas used for a transaction
+     */
+    function getL1GasUsed(bytes memory _data) public view returns (uint256);
+}
+```
+
+## Reserved System Address 1
+
+Reserved for future use.
+
+## Reserved System Address 2
+
+Reserved for future use.
+
+## L1Attributes
+
+[l1-attr-predeploy]: glossary.md#l1-attributes-predeployed-contract
+
+The [L1Attributes](l1-attr-predeploy) was introduced in bedrock and is responsible for
+mainting L1 context in L2. This allows for L1 state to be accessed in L2.
+
+```solidity
+interface L1Attributes {
+    /**
+     * @dev Returns the special account that can only send
+     * transactions to this contract
+     */
+    function DEPOSITOR_ACCOUNT() public returns (address);
+
+    /**
+     * @dev Returns the latest known L1 block number
+     */
+    function number() public returns (uint256);
+
+    /**
+     * @dev Returns the latest known L1 timestamp
+     */
+    function timestamp() public returns (uint256);
+
+    /**
+     * @dev Returns the latest known L1 basefee
+     */
+    function basefee() public returns (uint256);
+
+    /**
+     * @dev Returns the latest known L1 transaction hash
+     */
+    function hash() public returns (bytes32);
+
+    /**
+     * @dev sets the latest L1 block attributes
+     */
+    function setL1BlockValues(
+        uint256 _number,
+        uint256 _timestamp,
+        uint256 _basefee,
+        bytes32 _hash
+    ) external;
+}
+
+```
+
+## Withdrawer
+
+The `Withdrawer` is responsible for handling L2 to L1 interactions.
+
+```solidity
+/**
+ * @title Withdrawer
+ * @notice The Withdrawer contract facilitates sending both ETH value and data from L2 to L1.
+ * It is predeployed in the L2 state at address 0x4200000000000000000000000000000000000016.
+ */
+contract Withdrawer {
```

Comment:
nit: Can we embed contracts directly? My concern is them getting out of date. Another question is will we want to upgrade the pre-deploys? This is pretty out of scope and doesn't need to be addressed as part of this PR.

Reply:
The point here is that its all interfaces and that we can transparently change the predeploys in a backwards compatible way. I would much rather link to actual solidity files instead of embedding the source code in markdown, so I could add an `interfaces` directory if preferred

Reply:
Oh nice. That makes a bunch of sense (I just skimmed an saw solidity, not the fact that it was only interfaces).